text/0000-sync-networked-vars.md

```diff
@@ -0,0 +1,114 @@
+- Feature Name: `synced_networked_var`
+- Start Date: 2021-02-12
+- RFC PR: [Unity-Technologies/com.unity.multiplayer.rfcs#0000](https://github.com/Unity-Technologies/com.unity.multiplayer.rfcs/pull/0000)
+- Issue: [Unity-Technologies/com.unity.multiplayer#0000](https://github.com/Unity-Technologies/com.unity.multiplayer/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Make the NetworkedVariables synchronized with the tick system.
+
+We suggest two elements:
+- Changing `NetworkedBehaviour::NetworkedBehaviourUpdate()` so that the function runs at most one per Network Tick. When called more than once per Network Tick, it would simply return.
+- NetworkedVariables carrying two pieces of information with their delta updates:
+    - The local tick at which a variable was modified
+    - The remote tick at which the variable was modified, when the host relays the variable to clients, including the source client    
+
+# Motivation
+[motivation]: #motivation
+
+In a client-server topology, any client and the server can modify NetworkedVariables. The server receives the clients' changes and send them back to all of the clients. In order to retain generality, we'll assume that any variable can be modified by anyone and that the server can also modify the variable as they are received and sent. For example, we could have server-driven physics over client-driven character movement.
+
+It is proposed that the NetworkedVariable interpolation algorithm be easily customizable and that the user can chose whether it supports client or server side authoritative mechanics. To account for this, each NetworkedVaraible will keep track of the following:
+- When it was modified by a given client
+- When the server updated or transmitted it
+
+This will enable better client-side interpolation, lag compensation, and in the future provide the required flexibility for new systems (i.e. like snapshotting).
+
+This provides enough information such that the client can distinguish between its own updates from that of server-side adjustments (i.e. like physics).
+
+This will keep the overall bandwidth usage in check, preventing frequent updates for NetworkedVariables.
+
+This sets up MLAPI for upcoming features like snapshot.
+
+# Dependency
+[dependency]: #dependency
+
+This feature depends on a Tick System that would provide:
+- A discrete measurement of the local time, based on an integer number of tick elapsed
+- The tick number being updated at the very beginning of each frame and stable throughout the frame
+- A tick that is based on the real time elapsed since a fixed point in the past 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Some definitions:
+
+#### Local tick
+The tick number at which a variable was modified, relative to the tick system on this machine. If I get an update for a variable I changed, this tell me: when did I change it?
```

Comment:
I'm still coming around to why there are separate tick numbers for client and server vs. them all marching to the same tick number

Reply:
a common example: 60 ticks per second for client to server input rate & 20 ticks per second for server to client snapshot rate.

Reply:
I'd prefer the "60 ticks per second client with 20 ticks per second server" situation to result in the server increasing the tick by 3, three times less often, while the client updates the tick by 1, 60 times per second. But we can hash that out in the TickSystem RFC. :-)